### ISSUE TITLE
Incorporate line lists

### ISSUE THREAD

**Issue by kassin** [spacetelescope/specviz#291]
Tony Marston created a document with line lists and some ideas for incorporating them into SpecViz.

https://stsci.slack.com/files/anmarston/F4UR94WUX/jwst_line_list.pdf

**Comment by ibusko** [comment]
The link above is broken. Can't find the document.

**Comment by ibusko** [comment]
I have a question about the general behavior of this feature. When the user wants to display line identifications on top of whatever spectrum is being plotted, he/she has to perform an action, that right now is to click on the Line Labels green button on the top menu.

My question relates to what is to happen afterwards. One possible option is to present the user with the file selector widget, where he/she will pick up one file with the line list. Then specviz will create another GUI widget (could be a plugin) which will display the list contents in, say, tabular form. This widget will enable the user to select the specific lines to be displayed.

The other option (the one I was considering initially as it is based on software I wrote in the past) provides pre-canned line lists that are part of the specviz distribution itself. In this distribution, line lists such as the ones put together by Tony in his document, would be already present in the tool, hidden from the user. As soon as he/she clicks on the green Line Labels button, the specialized GUI widget pops up into existence with the line lists already displayed in tabular form. 

In this last option, we should eventually provide the ability for the user to input his/hers own line lists. That might be involved though, since there is no universal format for the lists and metadata information must be put together by the user for the software to be able to decode the list.

**Comment by hcferguson** [comment]
I think we will want to provide some pre-canned lists regardless of whether the user has to read them in or just click a "default" button. The ability to toggle them on and off with a hot-key would be useful. I do like the suggestion of popping up a table and allowing the user to select certain lines, although most useful there might be "show me all the hydrogen lines" or "show me only the strong lines" which might be rather tricky to implement (so out of scope for an initial implementation).

**Comment by ibusko** [comment]
Functionality such as "show me all the hydrogen lines" or "show me only the strong lines" can be implemented in a low-cost way by having the table to be sortable in each column. This ability exists in QTableView and the code I implemented already exploits it. For instance, one can sort the table on the line ID column and thus have all H lines collected together in a contiguous block of rows. Then one can just multiple-select those rows. It's a bit cumbersome, but I suppose any other approach would be cumbersome in some way.

**Comment by ibusko** [comment]
Should we then use the line lists provided by Tony Marston as the default set?

**Comment by kassin** [comment]
I agree with you both: having the option to show specific sets of lines (or selected lines) would be great. Yes,  please implement the lines from Tony Marston's list.  Also feel free to contact him about his thoughts on implementation.  His office is in the south corner of the 4th floor in Muller.

**Comment by kassin** [comment]
New files from Tony (line list document + text file with data):
https://stsci.box.com/s/26y950pp3etoojujx8pftxrcs2zu4qth
https://stsci.box.com/s/o56vm0ba186l498ed2947kta3k51jciq

**Comment by ibusko** [comment]
I've been working with Tony's lists; by now they are all included in the software. Most of the functionality he suggested in the document is also in place (it's still all in my personal github fork only). A question: these line lists are targeted to JWST needs and thus do not cover shorter wavelength regimes. Since specviz supports data from many sources, shouldn't we provide a couple of line lists that cover the UV-optical region as well? 

I had collected (for the old Java specview app) a number of them. Below I list pointers for the original data. The way I coded the line list handling in specviz supports these lists almost transparently, so it's a low cost add-on to add them to the package:

http://cdsweb.u-strasbg.fr/cgi-bin/qcat?VI/16
http://cdsweb.u-strasbg.fr/cgi-bin/qcat?VI/71A
http://cdsweb.u-strasbg.fr/cgi-bin/qcat?VI/84

These are lists (also included in the specview app) that are targeted to the IR, so they might be redundant at this point. But are easy to include in specviz if we wish so:

http://www.mpe.mpg.de/ir/ISO/linelists/FSlines.html
 http://www.mpe.mpg.de/ir/ISO/linelists/H2.html
http://www.mpe.mpg.de/ir/ISO/linelists/Molecular.html
http://www.mpe.mpg.de/ir/ISO/linelists/Hydrogenic.html
http://www.mpe.mpg.de/ir/ISO/linelists/CO.html

**Comment by apmarston** [comment]
Hi Ivo,
Regarding optical lines. These would also be useful for those observing significantly redshifted sources. So from a JWST perspective I think it would be good to add them.
The ISO lists you note at the end are already incorporated within the lists I passed on to you which you note you have already installed.
Cheers
Tony
